oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,393 @@
+    #[serde(rename = "desired_marker_positions")]
+    pub desired_marker_pos: [f64; 5],
+    /// The increment for the desired marker positions.
+    #[serde(rename = "desired_marker_increments")]
+    pub desired_marker_incrs: [f64; 5],
+}
+
+impl Quantile {
+    /// Create a new `Quantile` instance.
+    pub fn new(p: u64) -> Result<Self, QuantileError> {
+        let p = p as f64 / 100.;
+        if p < 0. || p > 1. {
+            return Err(QuantileError::InvalidPValue);
+        }
+
+        Ok(Self {
+            marker_hghts: [0.; 5],
+            marker_pos: [1, 2, 3, 4, 0],
+            desired_marker_pos: [1., 1. + 2. * p, 1. + 4. * p, 3. + 2. * p, 5.],
+            desired_marker_incrs: [0., p / 2., p, (1. + p) / 2., 1.],
+        })
+    }
+
+    /// Create a new `Quantile` instance from the given marker heights and
+    /// positions.
+    pub fn from(
+        marker_hghts: [f64; 5],
+        marker_pos: [i64; 5],
+        desired_marker_pos: [f64; 5],
+        desired_marker_incrs: [f64; 5],
+    ) -> Self {
+        Self {
+            marker_hghts,
+            marker_pos,
+            desired_marker_pos,
+            desired_marker_incrs,
+        }
+    }
+
+    /// Construct a `Quantile` instance for the 50th/median percentile.
+    pub fn p50() -> Self {
+        Self::new(50).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 90th percentile.
+    pub fn p90() -> Self {
+        Self::new(90).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 95th percentile.
+    pub fn p95() -> Self {
+        Self::new(95).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 99th percentile.
+    pub fn p99() -> Self {
+        Self::new(99).unwrap()
+    }
+
+    /// Get the p value as a float.
+    fn _p(&self) -> f64 {
+        self.desired_marker_incrs[2]
+    }
+
+    /// Get the p value as an integer.
+    pub fn p(&self) -> u64 {
+        (self.desired_marker_incrs[2] * 100.0) as u64
+    }
+
+    /// Estimate the p-quantile of the population.
+    ///
+    /// Returns an error if the sample is empty.
+    pub fn estimate(&self) -> Result<f64, QuantileError> {
+        // Return NaN if the sample is empty.
+        if self.is_empty() {
+            return Err(QuantileError::EmptyQuantile);
+        }
+
+        // Return the middle marker height if the sample size is at least 5.
+        if self.len() >= 5 {
+            return Ok(self.marker_hghts[2]);
+        }
+
+        let mut heights: [f64; 4] = [
+            self.marker_hghts[0],
+            self.marker_hghts[1],
+            self.marker_hghts[2],
+            self.marker_hghts[3],
+        ];
+
+        let len = self.len() as usize;
+        float_ord::sort(&mut heights[..len]);
+
+        let desired_index = (len as f64) * self._p() - 1.;
+        let mut index = desired_index.ceil();
+        if desired_index == index && index >= 0. {
+            let index = index.round_ties_even() as usize;
```

Comment:
I don't follow why we're using `round_ties_even()` rather than `round()`. Could you make a note explaining that?